cmd/autoretrieve/bitswap/provider.go

```diff
@@ -0,0 +1,210 @@
+)
+
+type ProviderConfig struct {
+	DataDir string
+}
+
+type Provider struct {
+	config        ProviderConfig
+	network       network.BitSwapNetwork
+	blockManager  *blocks.Manager
+	retriever     *filecoin.Retriever
+	blockQueues   map[peer.ID]BlockQueue
+	blockQueuesLk sync.Mutex
+}
+
+type BlockQueue struct {
+	wantHaveChan  chan blocks.Block
+	wantBlockChan chan blocks.Block
+}
+
+func NewProvider(
+	config ProviderConfig,
+	host host.Host,
+	datastore datastore.Batching,
+	blockManager *blocks.Manager,
+	retriever *filecoin.Retriever,
+) (*Provider, error) {
+
+	fullRT, err := fullrt.NewFullRT(host, dht.DefaultPrefix, fullrt.DHTOption(
+		dht.Datastore(datastore),
+		dht.BucketSize(20),
+		dht.BootstrapPeers(dht.GetDefaultBootstrapPeerAddrInfos()...),
+	))
+	if err != nil {
+		return nil, err
+	}
+
+	provider := &Provider{
+		config:       config,
+		network:      network.NewFromIpfsHost(host, fullRT),
+		blockManager: blockManager,
+		retriever:    retriever,
+		blockQueues:  make(map[peer.ID]BlockQueue),
+	}
+
+	provider.network.SetDelegate(provider)
+
+	return provider, nil
+}
+
+// Upon receiving a message, provider will iterate over the requested CIDs. For
+// each CID, it'll check if it's present in the blockstore. If it is, it will
+// respond with that block, and if it isn't, it'll start a retrieval and
+// register the block to be sent later.
+func (provider *Provider) ReceiveMessage(ctx context.Context, sender peer.ID, incoming message.BitSwapMessage) {
+
+	// For each item in the incoming wantlist, we either need to say whether we
+	// have the block or not, or send the actual block.
+
+entryLoop:
+	for _, entry := range incoming.Wantlist() {
+
+		// Only respond to WANT_HAVE and WANT_BLOCK
+		if entry.WantType != wantTypeHave && entry.WantType != wantTypeBlock {
+			continue
+		}
+
+		// First we check the local blockstore
+		block, err := provider.blockManager.Get(entry.Cid)
+
+		switch {
+		case err == nil:
+			// If we did have the block locally, just respond with that
+			provider.respondPositive(sender, entry, block)
+		default:
+			// Otherwise, we will need to ask for it...
+			if err := provider.retriever.Request(entry.Cid); err != nil {
+				// If request failed, it means there's no way we'll be able to
+				// get that block at the moment, so respond with DONT_HAVE
+				provider.respondNegative(sender, entry)
+				continue entryLoop
+			}
+
+			var queue chan blocks.Block
+
+			provider.blockQueuesLk.Lock()
+			switch entry.WantType {
+			case wantTypeHave:
+				queue = provider.blockQueues[sender].wantHaveChan
+			case wantTypeBlock:
+				queue = provider.blockQueues[sender].wantBlockChan
+			}
+			provider.blockQueuesLk.Unlock()
+
+			provider.blockManager.GetAwait(entry.Cid, queue)
+		}
+	}
+}
+
+func (provider *Provider) ReceiveError(err error) {
+	logger.Errorf("Error receiving bitswap message: %v", err)
+}
+
+func (provider *Provider) PeerConnected(peer peer.ID) {
+	blockQueue := BlockQueue{
```

Comment:
you should probably use the peertaskqueue for this.

The code lives here: https://github.com/ipfs/go-peertaskqueue

Basically, you make a queue, and then you push tasks onto it. So below here instead of doing the sends directly (which could block and gum things up elsewhere) you could push a task onto the queue with a topic of "send block", and the data could be the block (or maybe the data would be the cid, so you dont have to keep the blocks in memory for a while).

Then, you have other worker gorotuines calling "PopTasks" and executing the tasks.